doc/design/parallel-node-config.md

```diff
@@ -0,0 +1,214 @@
+DrainIdle                       = "Idle"
+DrainRequired                   = "Drain_Required"
+RebootRequired                  = "Reboot_Required"
+Draining                        = "Draining"
+DrainComplete                   = "DrainComplete"
+```
+
+Drain controller will watch for node annotation, `sriovnetwork.openshift.io/state` 
+and SriovNetworkNodeState annotation `sriovnetwork.openshift.io/desired-state`
+and write the `sriovnetwork.openshift.io/current-state` annotation in the SriovNetworkNodeState.
+
+Config Daemon will read SriovNetworkNodeState annotation `sriovnetwork.openshift.io/desired-state` and write both
+node annotation, `sriovnetwork.openshift.io/state` and SriovNetworkNodeState annotation `sriovnetwork.openshift.io/desired-state`.
+
+*NOTE:* In the future we are going to drop the node annotation and only use the SriovNetworkNodeState
+
+Draining procedure:
+
+1. config daemon mark the node as `Drain_Required` or `Reboot_Required`
+2. operator drain controller reconcile loop find the node SriovNetworkPoolConfig
+   1. if number of `Draining` nodes is great or equal to the `MaxUnavailable` the operator will re-queue the request
+   2. if number of `Draining` nodes is lower than the `MaxUnavailable` the operator will start the draining process
+5. on Openshift platform we will pause the machine config pool related to the node
+6. the operator will start the drain process
+   1. if `Drain_Required` the operator will remove ONLY pods used sriov devices
+   2. if `Reboot_Required` the operator will remove ALL the pods on the system
+9. operator moves the `sriovnetwork.openshift.io/current-state` annotation to `DrainComplete`
+10. daemon will continue to the configuration when it's done it will move back the `sriovnetwork.openshift.io/state` annotation to `Idle`
+11. operator runs the complete drain to remove the cordon and mark the `sriovnetwork.openshift.io/current-state` annotation to `Idle`
+
+### API Extensions
+
+#### Extend existing CR SriovNetworkPoolConfig
+SriovNetworkPoolConfig is used only for OpenShift to provide configuration for
+OVS Hardware Offloading. We can extend it to add configuration for the drain
+pool. E.g.:
+
+```golang
+// SriovNetworkPoolConfigSpec defines the desired state of SriovNetworkPoolConfig
+type SriovNetworkPoolConfigSpec struct {
+    ...
+	
+    // nodeSelector specifies a label selector for Nodes
+    NodeSelector *metav1.LabelSelector `json:"nodeSelector,omitempty"`
+
+    // maxUnavailable defines either an integer number or percentage
+    // of nodes in the pool that can go Unavailable during an update.
+    //
+    // A value larger than 1 will mean multiple nodes going unavailable during
+    // the update, which may affect your workload stress on the remaining nodes.
+    // Drain will respect Pod Disruption Budgets (PDBs) such as etcd quorum guards,
+    // even if maxUnavailable is greater than one.
+    MaxUnavailable *intstr.IntOrString `json:"maxUnavailable,omitempty"`
```

Comment:
this is relevant for drain/parallel configuration only right ?
we may have additional parameters down the road. 

maybe put this one under "DrainConfig" similarly to the existing "OvsHardwareOffloadConfig"

Reply:
you are right that is for draining but a configuration of a pool. and I must say I am not familiar with having `DrainConfig` with only field inside it just makes the API more complicated that he needs to be if you ask me :)

Reply:
what about :

> we may have additional parameters down the road.

Reply:
The thing is we are going to deprecate `OvsHardwareOffloadConfig` the pool CRD should never have contained that.
everything related to a pool of nodes should go under this CRD if we decide in the future to have priority or `skipDrainOnReboot` or maybe even the creation of the switches for switchdev part for example should be part of the CRD directly without another layer inside of it.
That is my take if you ask me

Reply:
as there is a potential to have additional configuration down the road, i think its worth "partitioning" according to domains now and not later.

if you feel strongly about keeping MaxUnavailable top level im OK.